Gemfile

```
@@ -1,9 +1,14 @@
 source "https://rubygems.org"
-
 gemspec
 
-gem "json"
+case redis_version = ENV.fetch('REDIS_VERSION', 'latest')
+when 'latest'
+  gem 'redis', '~> 4.7'
+else
+  gem 'redis', "~> #{redis_version}"
```

I think you'll need to change this, to avoid loading a higher than desired version.  E.g:
```suggestion
  gem 'redis', "~> #{redis_version}.0"
```

If you look at the CI logs, you can see that it isn't installing the requested major.minor version, but the latest (installable) major version.

That's in addition to adding some matrix exclusions for old rubies...